File: src/types/graph.ts
import type { SessionId } from './api'
import type { BookSourceRef } from './book'

export type GraphQueryRequest = {
  query: string
  session_id: SessionId
}

export type GraphNodeType =
  | 'artifact'
  | 'dynasty'
  | 'material'
  | 'category'
  | 'pattern'
  | 'site'
  | 'craft'
  | 'inscription'
  | 'usage'
  | 'concept'
  | 'person'
  | 'collection'
  | 'other'
  | (string & {})

export type GraphNode = {
  id: string
  label: string
  nodeType: GraphNodeType
  description?: string | null
  sourceRefs?: BookSourceRef[]
}

export type GraphEdge = {
  id: string
  source: string
  target: string
  label: string
  weight?: number
}

export type GraphResponse = {
  graphId: string
  centerNodeId: string
  nodes: GraphNode[]
  edges: GraphEdge[]
}

File: app/schemas/graph.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.common import SourceRef

GraphNodeType = Literal[
    "knowledge_base",
    "book",
    "chapter",
    "section",
    "markdown_note",
    "artifact",
    "person",
    "dynasty",
    "site",
    "concept",
    "material",
    "pattern",
    "craft",
    "category",
    "inscription",
    "usage",
    "other",
]


class GraphNode(BaseModel):
    id: str
    label: str
    nodeType: GraphNodeType
    description: str | None = None
    sourceRefs: list[SourceRef] | None = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    label: str
    weight: float | None = None
    sourceRefs: list[SourceRef] | None = None


class GraphResponse(BaseModel):
    graphId: str
    centerNodeId: str
    nodes: list[GraphNode]
    edges: list[GraphEdge]
